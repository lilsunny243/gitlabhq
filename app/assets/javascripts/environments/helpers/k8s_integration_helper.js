import { differenceInSeconds } from '~/lib/utils/datetime_utility';
import { CLUSTER_AGENT_ERROR_MESSAGES } from '../constants';

export function generateServicePortsString(ports) {
  if (!ports?.length) return '';

  return ports
    .map((port) => {
      const nodePort = port.nodePort ? `:${port.nodePort}` : '';
      return `${port.port}${nodePort}/${port.protocol}`;
    })
    .join(', ');
}

export function getServiceAge(creationTimestamp) {
  if (!creationTimestamp) return '';

  const timeDifference = differenceInSeconds(new Date(creationTimestamp), new Date());

  const seconds = Math.floor(timeDifference);
  const minutes = Math.floor(seconds / 60) % 60;
  const hours = Math.floor(seconds / 60 / 60) % 24;
  const days = Math.floor(seconds / 60 / 60 / 24);

  let ageString;
  if (days > 0) {
    ageString = `${days}d`;
  } else if (hours > 0) {
    ageString = `${hours}h`;
  } else if (minutes > 0) {
    ageString = `${minutes}m`;
  } else {
    ageString = `${seconds}s`;
  }

  return ageString;
}

export function getDeploymentsStatuses(items) {
  const failed = [];
  const ready = [];

  items.forEach((item) => {
    const [available, progressing] = item.status?.conditions ?? [];
    // eslint-disable-next-line @gitlab/require-i18n-strings
    if (available.status === 'True') {
      ready.push(item);
      // eslint-disable-next-line @gitlab/require-i18n-strings
    } else if (available.status !== 'True' && progressing.status !== 'True') {
      failed.push(item);
    }
  });

  return {
    ...(failed.length && { failed }),
    ...(ready.length && { ready }),
  };
}

export function getDaemonSetStatuses(items) {
  const failed = items.filter((item) => {
    return (
      item.status?.numberMisscheduled > 0 ||
      item.status?.numberReady !== item.status?.desiredNumberScheduled
    );
  });
  const ready = items.filter((item) => {
    return (
      item.status?.numberReady === item.status?.desiredNumberScheduled &&
      !item.status?.numberMisscheduled
    );
  });

  return {
    ...(failed.length && { failed }),
    ...(ready.length && { ready }),
  };
}

export function getStatefulSetStatuses(items) {
  const failed = items.filter((item) => {
    return item.status?.readyReplicas < item.spec?.replicas;
  });
  const ready = items.filter((item) => {
    return item.status?.readyReplicas === item.spec?.replicas;
  });

  return {
    ...(failed.length && { failed }),
    ...(ready.length && { ready }),
  };
}

export function getReplicaSetStatuses(items) {
  const failed = items.filter((item) => {
    return item.status?.readyReplicas < item.spec?.replicas;
  });
  const ready = items.filter((item) => {
    return item.status?.readyReplicas === item.spec?.replicas;
  });

  return {
    ...(failed.length && { failed }),
    ...(ready.length && { ready }),
  };
}

export function getJobsStatuses(items) {
  const failed = items.filter((item) => {
    return item.status.failed > 0 || item.status?.succeeded !== item.spec?.completions;
  });
  const completed = items.filter((item) => {
    return item.status?.succeeded === item.spec?.completions;
  });

  return {
    ...(failed.length && { failed }),
    ...(completed.length && { completed }),
  };
}

export function getCronJobsStatuses(items) {
  const failed = [];
  const ready = [];
  const suspended = [];

  items.forEach((item) => {
    if (item.status?.active > 0 && !item.status?.lastScheduleTime) {
      failed.push(item);
    } else if (item.spec?.suspend) {
      suspended.push(item);
    } else if (item.status?.lastScheduleTime) {
      ready.push(item);
    }
  });

  return {
    ...(failed.length && { failed }),
    ...(suspended.length && { suspended }),
    ...(ready.length && { ready }),
  };
}

export function humanizeClusterErrors(reason) {
  const errorReason = String(reason).toLowerCase();
  const errorMessage = CLUSTER_AGENT_ERROR_MESSAGES[errorReason];
  return errorMessage || CLUSTER_AGENT_ERROR_MESSAGES.other;
}
