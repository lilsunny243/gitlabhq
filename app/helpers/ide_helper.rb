# frozen_string_literal: true

module IdeHelper
  def ide_data(project:, branch:, path:, merge_request:, fork_info:, learn_gitlab_source:)
    {
      'can-use-new-web-ide' => can_use_new_web_ide?.to_s,
      'use-new-web-ide' => use_new_web_ide?.to_s,
      'new-web-ide-help-page-path' => help_page_path('user/project/web_ide/index.md', anchor: 'vscode-reimplementation'),
      'user-preferences-path' => profile_preferences_path,
      'branch-name' => branch,
      'file-path' => path,
      'fork-info' => fork_info&.to_json,
      'editor-font-src-url' => font_url('jetbrains-mono/JetBrainsMono.woff2'),
      'editor-font-family' => 'JetBrains Mono',
      'editor-font-format' => 'woff2',
      'merge-request' => merge_request,
      'learn-gitlab-source' => (!!learn_gitlab_source).to_s
    }.merge(use_new_web_ide? ? new_ide_data(project: project) : legacy_ide_data(project: project))
  end

  def can_use_new_web_ide?
    Feature.enabled?(:vscode_web_ide, current_user)
  end

  def use_new_web_ide?
    can_use_new_web_ide? && !current_user.use_legacy_web_ide
  end

  private

  def new_ide_data(project:)
    {
      'project-path' => project&.path_with_namespace,
      'csp-nonce' => content_security_policy_nonce,
      # We will replace these placeholders in the FE
      'ide-remote-path' => ide_remote_path(remote_host: ':remote_host', remote_path: ':remote_path')
    }
  end

  def legacy_ide_data(project:)
    {
      'empty-state-svg-path' => image_path('illustrations/multi_file_editor_empty.svg'),
      'no-changes-state-svg-path' => image_path('illustrations/multi-editor_no_changes_empty.svg'),
      'committed-state-svg-path' => image_path('illustrations/multi-editor_all_changes_committed_empty.svg'),
      'pipelines-empty-state-svg-path': image_path('illustrations/pipelines_empty.svg'),
      'switch-editor-svg-path': image_path('illustrations/rocket-launch-md.svg'),
      'promotion-svg-path': image_path('illustrations/web-ide_promotion.svg'),
      'ci-help-page-path' => help_page_path('ci/quick_start/index'),
      'web-ide-help-page-path' => help_page_path('user/project/web_ide/index.md'),
      'render-whitespace-in-code': current_user.render_whitespace_in_code.to_s,
      'default-branch' => project && project.default_branch,
      'project' => convert_to_project_entity_json(project),
      'enable-environments-guidance' => enable_environments_guidance?(project).to_s,
      'preview-markdown-path' => project && preview_markdown_path(project),
      'web-terminal-svg-path' => image_path('illustrations/web-ide_promotion.svg'),
      'web-terminal-help-path' => help_page_path('user/project/web_ide/index.md', anchor: 'interactive-web-terminals-for-the-web-ide'),
      'web-terminal-config-help-path' => help_page_path('user/project/web_ide/index.md', anchor: 'web-ide-configuration-file'),
      'web-terminal-runners-help-path' => help_page_path('user/project/web_ide/index.md', anchor: 'runner-configuration')
    }
  end

  def convert_to_project_entity_json(project)
    return unless project

    API::Entities::Project.represent(project, current_user: current_user).to_json
  end

  def enable_environments_guidance?(project)
    experiment(:in_product_guidance_environments_webide, project: project) do |e|
      e.candidate { !has_dismissed_ide_environments_callout? }

      e.run
    end
  end

  def has_dismissed_ide_environments_callout?
    current_user.dismissed_callout?(feature_name: 'web_ide_ci_environments_guidance')
  end
end
