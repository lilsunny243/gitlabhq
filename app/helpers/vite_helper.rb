# frozen_string_literal: true

module ViteHelper
  def universal_javascript_include_tag(*args)
    if vite_enabled
      vite_javascript_tag(*args)
    else
      javascript_include_tag(*args)
    end
  end

  def universal_asset_path(*args)
    if vite_enabled
      vite_asset_path(*args)
    else
      asset_path(*args)
    end
  end

  private

  def vite_enabled
    Feature.enabled?(:vite) && !Rails.env.test? && vite_running
  end

  def vite_running
    ViteRuby.instance.dev_server_running?
  end
end
