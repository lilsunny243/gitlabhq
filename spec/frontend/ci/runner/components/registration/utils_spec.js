import { TEST_HOST } from 'helpers/test_constants';
import {
  DEFAULT_PLATFORM,
  LINUX_PLATFORM,
  MACOS_PLATFORM,
  WINDOWS_PLATFORM,
} from '~/ci/runner/constants';

import {
  commandPrompt,
  registerCommand,
  runCommand,
} from '~/ci/runner/components/registration/utils';

const REGISTRATION_TOKEN = 'REGISTRATION_TOKEN';
const DUMMY_GON = {
  gitlab_url: TEST_HOST,
};

describe('registration utils', () => {
  let originalGon;

  beforeAll(() => {
    originalGon = window.gon;
    window.gon = { ...DUMMY_GON };
  });

  afterAll(() => {
    window.gon = originalGon;
  });

  describe.each([DEFAULT_PLATFORM, LINUX_PLATFORM, MACOS_PLATFORM, WINDOWS_PLATFORM, null])(
    'for "%s" platform',
    (platform) => {
      describe('commandPrompt', () => {
        it('matches snapshot', () => {
          expect(commandPrompt({ platform })).toMatchSnapshot();
        });
      });
      describe('registerCommand', () => {
        it('matches snapshot', () => {
          expect(
            registerCommand({ platform, registrationToken: REGISTRATION_TOKEN }),
          ).toMatchSnapshot();
        });
      });
      describe('runCommand', () => {
        it('matches snapshot', () => {
          expect(runCommand({ platform })).toMatchSnapshot();
        });
      });
    },
  );
});
