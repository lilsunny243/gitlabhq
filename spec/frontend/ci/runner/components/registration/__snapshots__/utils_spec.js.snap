// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration utils for "linux" platform commandPrompt matches snapshot 1`] = `"$"`;

exports[`registration utils for "linux" platform commandPrompt matches snapshot 2`] = `"$"`;

exports[`registration utils for "linux" platform registerCommand matches snapshot 1`] = `
Array [
  "gitlab-runner register",
  "  --url http://test.host",
  "  --registration-token REGISTRATION_TOKEN",
]
`;

exports[`registration utils for "linux" platform registerCommand matches snapshot 2`] = `
Array [
  "gitlab-runner register",
  "  --url http://test.host",
  "  --registration-token REGISTRATION_TOKEN",
]
`;

exports[`registration utils for "linux" platform runCommand matches snapshot 1`] = `"gitlab-runner run"`;

exports[`registration utils for "linux" platform runCommand matches snapshot 2`] = `"gitlab-runner run"`;

exports[`registration utils for "null" platform commandPrompt matches snapshot 1`] = `"$"`;

exports[`registration utils for "null" platform registerCommand matches snapshot 1`] = `
Array [
  "gitlab-runner register",
  "  --url http://test.host",
  "  --registration-token REGISTRATION_TOKEN",
]
`;

exports[`registration utils for "null" platform runCommand matches snapshot 1`] = `"gitlab-runner run"`;

exports[`registration utils for "osx" platform commandPrompt matches snapshot 1`] = `"$"`;

exports[`registration utils for "osx" platform registerCommand matches snapshot 1`] = `
Array [
  "gitlab-runner register",
  "  --url http://test.host",
  "  --registration-token REGISTRATION_TOKEN",
]
`;

exports[`registration utils for "osx" platform runCommand matches snapshot 1`] = `"gitlab-runner run"`;

exports[`registration utils for "windows" platform commandPrompt matches snapshot 1`] = `">"`;

exports[`registration utils for "windows" platform registerCommand matches snapshot 1`] = `
Array [
  ".\\\\gitlab-runner.exe register",
  "  --url http://test.host",
  "  --registration-token REGISTRATION_TOKEN",
]
`;

exports[`registration utils for "windows" platform runCommand matches snapshot 1`] = `".\\\\gitlab-runner.exe run"`;
